Update signup progress state to a keyed object

spen opened the pull request:
> ~Note: This should be landed after #34770 is landed to avoid conflicts~

#### Changes proposed in this Pull Request

* Convert the signup progress store from an ordered array to an object, keyed by the names of the steps progressed.
* Access the relevant progress state according to the current flow.

#### Some notes:

Rather than the progress state being an exact representation of the steps currently progressed in the current flow, this approach attempts to make the progress store a collection of any steps that have been progressed during the signup journey.
Currently, when we switch flow we have to run logic on the progress store in an attempt to remove steps that are no longer needed, which is hard to do accurately and is causing bugs. It also means that we'd eagerly get rid of a step's progress if it didn't apply to the incoming flow, though it might become relevant again if the customer were to change their mind and switch back to that original flow. In this new approach we would hold on to the progress of every step that the customer has seen, regardless of its present usefulness and wouldn't need to attempt to keep the progress in sync with the flows.
In order to display the `3 out of 6 steps` information we just use the list of step names provided by the flow definition and match up the relevant steps found in the progress store. 

There are a few more notes over at #34885 (superseded by this PR) 

#### Testing instructions

**Testing user step last A/B test behaviour**
- Please follow the test instructions found in #34770 - whether that PR is landed yet the behaviour should remain the same as described there.
- More test instructions to come soon...




Fixes #32506

spen committed:
Update signup progress state to a keyed object

adamziel commented:
Just noting there are some conflicts and failing tests

spen commented:
Thanks Adam, I'm hoping to get some 👀 on this approach before potentially wasting time fixing/writing tests if the team don't want to move in this direction, or if there's something fundamentally wrong with this approach 👍
Same goes for running on the rebase treadmill, I don't want to get in to that loop until we're confident we want to merge.

p-jackson commented:
@spen fwiw I much prefer this approach (granted I've mainly looked at #35139). I'm not sure about calling it `progress` anymore though. That word does seem like a linear thing, so it makes me think it would be an array.

Would `stepData` or `stepState` be more accurate?

Also given so many components still expect the data in array form (and so there's many calls to `getFilteredSteps()`) maybe there should be a new selector for transforming the data into the shape expected by components? That way it's memoizable etc. etc., all the standard Redux tricks.

spen commented:
> I'm not sure about calling it progress anymore though... Would stepData or stepState be more accurate?

Totally! I've been having similar thoughts too! To me, simply `signup.steps` would be ideal.
I think that at some point we should take a more holistic look at the state and figure out what should go where, what the correct semantics and what we can get rid of to simplify our state and make it more reliable.

> maybe there should be a new selector for transforming the data into the shape expected by components? That way it's memoizable etc. etc., all the standard Redux tricks.

Yeah! I'm surprised this didn't cross my mind to be honest, it felt very icky making so many calls to the filtering functions in components.

Now that I know this is a direction we're considering I'll put some extra work in to neaten this up! 👌

spen committed:
Fix flow-controller tests

p-jackson commented:
> To me, simply `signup.steps` would be ideal

Yeah definitely

spen committed:
Update unit tests to match new data structure

adamziel commented on client/state/test/initial-state.js:
```diff
@@ -383,12 +383,12 @@ describe( 'initial-state', () => {
-						progress: [
-							{
+						progress: {
+							'logged-out-step': {
```
It would be nice to avoid duplication here - the key is the same as stepName. Let's not delay this PR though - just keep not of that for the future.

spen replied:
Totally! I'll look in to what it would take to do that soon after :)

adamziel reviewed (dismissed):
Did you test this PR @spen ? The back button alternates between the previous and the current step:

https://cld.wthms.co/DTIqg8

~~Edit: This seems to be on me - I'm experiencing some quirks in my local calypso setup.~~
Edit 2: That's actually a problem with this PR

adamziel commented on client/signup/utils.js:
```diff
@@ -177,7 +177,13 @@ export function getDesignTypeForSiteGoals( siteGoals, flow ) {
+		// filter steps...
+		filter( progress, step => includes( flow.steps, step.stepName ) ),
+		// then order according to the flow definition...
+		( { stepName } ) => flow.steps.indexOf( stepName )
```
What would happen if we started our journey in flow A but now we're in flow B?

spen replied:
Ahh well, if they have completed 3 steps from flow A, then switched to flow B and flow B shared 2 out the 3 steps from flow A, those 2 steps would be held in state (as is the 1 unshared step, actually) so when we call `getFilteredSteps` after switching to flow B 2 of it's steps will returned .

adamziel committed:
Fix the back button circling between two steps

adamziel commented:
I pushed a fix for the back button problem, will keep on testing

adamziel commented on client/state/signup/progress/reducer.js:
```diff
@@ -3,132 +3,120 @@
 
-	return addStep( state, { ...step, status: 'in-progress' } );
-}
+	return has( state, step.stepName )
```
Nitpick: Using lodash for simple stuff like that adds an extra step of "better look it up in lodash docs" while reviewing. `step.stepName in state` accomplishes the same thing and is easier to read

spen replied:
I'll leave this one be, this suggestion would error in the case where `state` is `undefined`.
I know thats hugely unlikely, but I always prefer to be consistent with patterns like this.
Also, I'd argue that any dev has to look up documentation when touching new code and we lodash is well used throughout Calypso, so the learning opportunity wouldn't hurt and could encourage others to use safer methods.

adamziel commented on client/state/signup/progress/reducer.js:
```diff
@@ -3,132 +3,120 @@
+
+const updateStep = ( state, newStepState ) => {
+	const { stepName, status } = newStepState;
+	const stepState = get( state, stepName );
```
Same lodash nitpick as in other comment: `state[stepName]` does the same and is simpler. I've seen some more examples but you get the idea.

adamziel commented on client/state/signup/progress/reducer.js:
```diff
@@ -3,132 +3,120 @@
+		};
+	}
 
-function removeUnneededSteps( state, { flowName } ) {
```
So glad to see this gone

adamziel commented on client/state/signup/progress/reducer.js:
```diff
@@ -3,132 +3,120 @@
 		[ SIGNUP_PROGRESS_COMPLETE_STEP ]: completeStep,
 		[ SIGNUP_PROGRESS_INVALIDATE_STEP ]: invalidateStep,
 		[ SIGNUP_PROGRESS_PROCESS_STEP ]: processStep,
-		[ SIGNUP_PROGRESS_REMOVE_UNNEEDED_STEPS ]: removeUnneededSteps,
```
This PR removes the reducer, but the constant and the action itself are still used in a few places. There are no other reducers handling this so it should be safe to remove any other references as well.

adamziel replied:
By the way it's not a blocker for merging - it can totally be adressed in a subsequent PR

spen replied:
Thanks for spotting this Adam! I've opened a follow up PR here: #35461
Once I finish up a test plan and once this is merged I'll publish as a proper PR :)

adamziel commented on client/signup/navigation-link/test/index.jsx:
```diff
@@ -77,6 +80,11 @@ describe( 'NavigationLink', () => {
 	test( 'should set a proper url as href prop when the direction is "back".', () => {
 		expect( getStepUrl ).not.toHaveBeenCalled();
 
+		getFilteredSteps.mockReturnValue( {
```
getFilteredSteps returns an array, not an object - thus this tests for the wrong type of value

adamziel replied:
Also it would be great to have some tests for getFilteredSteps itself - this PR seems to revolve around it, but it has no coverage

spen replied:
Good point, I'll add some today, thanks @adamziel !

spen replied:
Thanks for spotting that! - `getFilteredSteps` should now be pretty well covered in https://github.com/Automattic/wp-calypso/pull/35049/commits/def07a343be80fb790ad6c773eb35c5bd991346b

adamziel reviewed (dismissed):
I left some notes, but other than fixing tests I don't see any major blockers. I'd love to get more eyes on this PR and if another person says it works and doesn't introduce any problems - I think it's good to go!

As a side note, I think this even fixes this bug: https://github.com/Automattic/wp-calypso/issues/35363

p-jackson commented on client/signup/navigation-link/index.jsx:
```diff
@@ -51,14 +52,11 @@ export class NavigationLink extends Component {
-		);
+		let steps = getFilteredSteps( flowName, signupProgress );
+		steps = steps.slice( 0, findIndex( steps, step => step.stepName === stepName ) );
+		const previousStep = findLast( steps, step => ! step.wasSkipped && step.stepName !== stepName );
```
nbd, just a thought

adamziel replied:
@p-jackson ah yes, that was me not adjusting line 59 after introducing line 58 - good spot!

p-jackson reviewed (commented):
I agree with Adam's comments too, but in general this looks good and I far prefer this store layout. Proceeding to some manual testing.

p-jackson reviewed (commented):
I agree with Adam's comments too, but in general this looks good and I far prefer this store layout. Proceeding to some manual testing.

p-jackson commented:
Tests well for me. Also went through the cases in https://github.com/Automattic/wp-calypso/pull/34211#issuecomment-506298663 👍

spen commented on client/blocks/login/index.jsx:
```diff
@@ -146,7 +146,7 @@ class Login extends Component {
 			} );
 		} else {
-			if ( isArray( this.props.signupProgress ) ) {
+			if ( size( this.props.signupProgress ) ) {
```
Good call and I do actually agree, I'll fix this up :)

spen replied:
Done in https://github.com/Automattic/wp-calypso/pull/35049/commits/00edebe1d64019fb48d2c1d0f4781427a5d2e550 :)

spen committed:
Fix getFilteredSteps test errors

spen committed:
Add unit test coverage for getFilteredSteps

spen committed:
Use isEmpty vs size for better reasoning

adamziel commented on client/signup/navigation-link/index.jsx:
```diff
@@ -53,10 +53,17 @@ export class NavigationLink extends Component {
-		steps = steps.slice( 0, findIndex( steps, step => step.stepName === stepName ) );
-		const previousStep = findLast( steps, step => ! step.wasSkipped && step.stepName !== stepName );
+		if ( ! isArray( steps ) ) {
+			return { stepName: null };
```
Why would that not return an array though? It seems like this check would conceal a legitimate bug if it were to happen one day

spen replied:
Yeah, we could have ended up returning something else from `getFilteredSteps` and have been returning `{ stepName: null }` constantly which would have been tricky to debug - good catch!
I've removed this check 👍

adamziel commented on client/signup/navigation-link/index.jsx:
```diff
@@ -53,10 +53,17 @@ export class NavigationLink extends Component {
+		const previousSteps = steps.slice( 0, findIndex( steps, step => step.stepName === stepName ) );
+		const previousStep = findLast(
+			previousSteps,
+			step => ! step.wasSkipped && step.stepName !== stepName
```
As @p-jackson mentioned, we could remove `step.stepName !== stepName` as that's accounted for in the `slice` in the previous line

spen replied:
Taken care of in https://github.com/Automattic/wp-calypso/pull/35049/commits/b569dab205e97c98d816ee251ab1c5b9bd697620

spen committed:
Remvoe redundant check in findLast

p-jackson committed:
Mock getFilteredSteps to fix NavigationLink tests

NavigationLink tests failed depending on the order they were run in.
Giving getFilteredSteps a default implementation (of not filtering
anything out) means tests don't depend on each others mocking.

p-jackson commented on client/signup/test/utils.js:
```diff
@@ -72,6 +73,73 @@ describe( 'utils', () => {
+						expect( Array.isArray( result ) ).toBe( true );
+					} );
+
+					test( 'it should return only the step objects that match the flow', () => {
```
Personally prefer writing these sort of object/array tests using `toEqual` which does a deep equality check.
```javascript
expect( result ).toEqual( [ userStep, siteTypeStep ] );
```

Using `toContain` has better semantics, which I like, but using `toEqual` will also test:
- elements are in the correct order
- there aren't other random elements in there
- doubles as an `isArray` test too :)

p-jackson reviewed (commented):
Changes look good 👍
I pushed a wee fix to the `<NavigationLink>` unit tests; mocking meant they weren't running independently.

Failing canary test is signup related so I'll take a peek to see how real the issue is.

p-jackson commented:
Yup, e2e failures are real. Simplified steps for reproducing bug:

1. Start flow that pre-selects a plan (e.g. `/start/personal`)
2. Go through steps, any site type, any domain
3. Gets stuck on the "Give us one minute and we’ll move right along" animation forever.

ramonjd commented:
> Yup, e2e failures are real. 

Could it have anything to do with the pre-selected plan issue I mentioned in #35139 
?


https://github.com/Automattic/wp-calypso/pull/35139#discussion_r311359757

andrewserong commented on client/lib/signup/flow-controller.js:
```diff
@@ -100,6 +100,8 @@ assign( SignupFlowController.prototype, {
 	_resetStoresIfUserHasLoggedIn: function() {
 		if (
+			// TODO: Address this properly if we move forward with the user step being last
+			this._flowName !== 'onboarding-user-last' &&
```
The A/B test to move the user step has wrapped up (https://github.com/Automattic/wp-calypso/pull/35552), so this line can be removed now.

andrewserong commented on client/lib/signup/test/flow-controller.js:
```diff
@@ -85,7 +67,7 @@ describe( 'flow-controller', () => {
 			} );
 
-			expect( getSignupProgress( store.getState() ) ).toHaveLength( 0 );
+			expect( isEmpty( getSignupProgress( store.getState() ) ) ).toBe( true );
```
This is just a tiny style nit, but It looks like a common pattern in other tests in the codebase for checking for an empty object is to use Jest's `.toEqual( {} );` so the lodash `isEmpty` might be unnecessary here and on line 86.

andrewserong commented on client/state/signup/progress/reducer.js:
```diff
@@ -3,132 +3,120 @@
+		? updateStep( state, step )
+		: addStep( state, { ...step, status: 'in-progress' } );
 
 function setResumeAfterLogin( state, { resumeStep } ) {
```
Feel free to disregard this one, but it looks like you've refactored all of the other regular functions in this reducer to use arrow functions assigned to a `const`, but this one's still a regular function, so it stood out to me.

spen replied:
Thanks @andrewserong, I'll leave this one be for now though to avoid growing the diff.

p-jackson commented:
Yup, @ramonjd's hint in https://github.com/Automattic/wp-calypso/pull/35139#discussion_r311359757 was just what I needed. Pending excluded steps now working again.

I'll try rebasing this to remove the reference to the `moveUserStepPosition` @andrewserong pointed out.

p-jackson commented:
I think this PR is all working now. I haven't addressed any of the minor points from the reviews, I'll let you take a look @spen before we 🤞 move to the next step!

spen commented:
Awesome, thanks for sorting this @p-jackson ! I was stuck on the same issue yesterday too - I'd gone down a different route that wasn't so fruitful - the changes in https://github.com/Automattic/wp-calypso/pull/35049/commits/943ffdc89fd993180981d169e71745f9173b6a2a do a great job of solving this issue 👌

spen committed:
Use .toEqual( {} ) to check the progress is empty

spen committed:
Use proper equality check in favour of toContain etc

adamziel commented:
@andrewserong @ramonjd would you mind giving this PR a spin and merging if it's all good now? It's an important piece of code as it will help Samus and Luna build Signup flows and AB tests with confidence.

ramonjd commented:
Any particular flows we should test? I was going to run through site creation, jetpack connection... the usual suspects. 

Now that https://github.com/Automattic/wp-calypso/pull/34770 has been reverted and closed, we can safely ignore this PR's test description, right?

I'm still seeing the `moveUserStepPosition` AB test in this PR... going to rebase. 👍

p-jackson commented:
@ramonjd no specific flows, what you mentioned sounds good. Things that interact with dependencies in weird and wonderful ways are probably the highest risks e.g. adding query parameters to flows, anything that auto-submits a step.

p-jackson commented:
And yes these changes don't really have anything to do with #34770 anymore, so I'd disregard those parts.

ramonjd committed:
Fix flows with pre-selected plans

The flow controller wasn't processing pending steps that had been
excluded, like when a step pre-selects a plan. Issue was that the
processing code switched to using `getFilteredSteps` which skips
excluded steps. Going back to the old (pre-PR) code fixed the issue.

I think `getFilteredSteps` was used because it returns an array, however
the old code still works because `_.filter` always returns an array,
even when given an object. Reverting @p-jackson's work in 943ffdc89fd993180981d169e71745f9173b6a2a I accidentally overwrote in a force push :(

ramonjd commented:
Been through the following flows (successfully) so far....

- [ ] Login and create a site with onboarding flow
- [ ] Navigate between steps, changing between Business and Ecommerce site types, create an E-commerce step
- [ ] Go to /business and /premium flows, navigate between between Business and Ecommerce site types, select Professional and continue onto purchase
- [ ] art.blog: Create a site by clicking on CTA /start/user?site_type=blog&vertical=art&ref=dotblog-landing
- [ ] Try other permutations, e.g., /start/user?site_type=business&vertical=travel

andrewserong reviewed (approved):
Great work @spen — I've tested a number of different permutations of creating a site today, and haven't run into any issues. Looks like the failing e2e tests right now are Gutenberg related and not this PR, so personally I think this looks good to 🚢

I noticed that the bug https://github.com/Automattic/wp-calypso/issues/35363 is still present in this PR, but that isn't at all a reason to hold off on merging, and we can pursue a fix for that separately.

ramonjd commented:
Also checked Jetpack connection flow. State checks out:

<img width="837" alt="Screen Shot 2019-08-21 at 7 59 05 pm" src="https://user-images.githubusercontent.com/6458278/63422702-3825b800-c44e-11e9-8fe4-5facb46c1f57.png">

Good work!
